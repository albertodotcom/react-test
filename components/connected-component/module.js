
const INCREMENT = 'counter/INCREMENT';
export const increment = () => ({
  type: INCREMENT,
});

const DECREMENT = 'counter/DECREMENT';
export const decrement = () => ({
  type: DECREMENT,
});

export const counterReducer = (state = 0, action) => {
  switch (action.type) {
    case INCREMENT:
      return state + 1;
    case DECREMENT:
      return state - 1;
    default:
      return state;
  }
};

export const getCounterValue = state => state.counter;
