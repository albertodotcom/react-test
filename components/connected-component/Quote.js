import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { getQuote, fetchQuote } from './module';

const Quote = ({ quote, onFetchQuote }) => (
  <div>
    <button className="t-fetch-quote" onClick={onFetchQuote}>Show a quote</button>
    {quote.loading ? <p className="t-quote">Loading</p> : null}
    {
      quote.quote
        ? <quote className="t-quote">{quote.quote}, {quote.author}</quote>
        : null
    }
  </div>
);

Quote.propTypes = {
  quote: PropTypes.shape({
    quote: PropTypes.string,
    author: PropTypes.string,
    loading: PropTypes.bool.isRequired,
  }).isRequired,
  onFetchQuote: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  quote: getQuote(state),
});

// Sadly I didn't find a way to use the shorthand notation and mock the module
const mapDispatchToProps = dispatch => ({
  onFetchQuote: () => dispatch(fetchQuote()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Quote);
