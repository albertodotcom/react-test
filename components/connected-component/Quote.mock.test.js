/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';

import Quote from './Quote';
import * as QuoteModule from './module';

jest.mock('./module');

describe('Quote', () => {
  let store;

  beforeEach(() => {
    QuoteModule.getQuote.mockImplementation(() => ({ loading: false }));
    store = configureMockStore([])({});
  });

  // Test that the state is passed to the component
  it('gets values from the store', () => {
    const storeState = 'a test state';
    store.getState = jest.fn().mockReturnValue(storeState);

    shallow(<Quote store={store} />);

    expect(QuoteModule.getQuote).toBeCalledWith(storeState);
  });

  it('displays a loading text', () => {
    QuoteModule.getQuote.mockImplementation(() => ({
      loading: true,
    }));

    const wrapper = shallow(<Quote store={store} />).shallow();

    expect(wrapper.find('.t-quote').text()).toBe('Loading');
  });

  it('displays a quote', () => {
    QuoteModule.getQuote.mockImplementation(() => ({
      quote: 'foo',
      author: 'baz',
      loading: false,
    }));

    const wrapper = shallow(<Quote store={store} />).shallow();

    expect(wrapper.find('.t-quote').text()).toBe('foo, baz');
  });

  // Test that the action creator is connected to the component
  it('calls fetchQuote when the "Get Quote" button is clicked', () => {
    const testFetchAction = { type: 'TEST_FETCH' };
    QuoteModule.fetchQuote.mockImplementation(() => testFetchAction);

    const wrapper = shallow(<Quote store={store} />).shallow();
    wrapper.find('.t-fetch-quote').simulate('click');

    expect(QuoteModule.fetchQuote).toHaveBeenCalledTimes(1);
    expect(QuoteModule.fetchQuote).toBeCalledWith();
    expect(store.getActions()).toEqual([testFetchAction]);
  });

  describe('journey with snapshot', () => {
    it('loads and renders a quote', () => {
      QuoteModule.getQuote
      .mockImplementationOnce(() => ({
        loading: false,
      }))
      .mockImplementationOnce(() => ({
        loading: true,
      }))
      .mockImplementationOnce(() => ({
        quote: 'foo',
        author: 'baz',
        loading: false,
      }));
      QuoteModule.fetchQuote.mockImplementation(() => ({ type: 'TEST_JOURNEY' }));

      const wrapper = shallow(<Quote store={store} />).shallow();
      expect(wrapper.children()).toMatchSnapshot();

      wrapper.find('.t-fetch-quote').simulate('click');
      expect(wrapper.children()).toMatchSnapshot();
      expect(wrapper.children()).toMatchSnapshot();

      expect(QuoteModule.fetchQuote).toHaveBeenCalledTimes(1);
    });
  });
});
