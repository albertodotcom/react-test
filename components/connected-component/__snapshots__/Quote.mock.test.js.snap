// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quote journey with snapshot loads and renders a quote 1`] = `
<div
  className="buttons"
>
  <button
    className="t-fetch-quote"
    onClick={[Function]}
  >
    Show a quote
  </button>
</div>
`;

exports[`Quote journey with snapshot loads and renders a quote 2`] = `
<div
  className="buttons"
>
  <button
    className="t-fetch-quote"
    onClick={[Function]}
  >
    Show a quote
  </button>
</div>
`;

exports[`Quote journey with snapshot loads and renders a quote 3`] = `
<div
  className="buttons"
>
  <button
    className="t-fetch-quote"
    onClick={[Function]}
  >
    Show a quote
  </button>
</div>
`;
